Map to onnx tensor type and define Float16

mingmingtasd opened the pull request:
1. Remove the unnecessary parameter ` OperandDataType data_type` of  `CreateInitializer` method, map the data type to onnx tensor type
2. Define `class Float16` to optimize code
3.  Add a helper method `CreateScalarInitializer` to create scalar with empty shape

PTAL, thanks!

shiyi9801 reviewed (approved):
LGTM!

huningxin commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -673,15 +703,12 @@ GraphBuilderOrt::AddInstanceNormalizationOperation(
-            ConvertFloat32ToFloat16(scale_data);
-        scale_name = CreateInitializer<uint16_t>(constant_dims, scale_data_fp16,
-                                                 input_data_type);
+        std::vector<Float16> scale_data_fp16 = Float16FromFloat32(scale_data);
```
Construct `scale_data_fp16` directly? I know this is an old issue, would be great to fix in this PR: 
```suggestion
        std::vector<Float16> scale_data_fp16(input_channel, Float16FromFloat32(1.0f));
```

huningxin commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -699,15 +726,12 @@ GraphBuilderOrt::AddInstanceNormalizationOperation(
-            ConvertFloat32ToFloat16(bias_data);
-        bias_name = CreateInitializer<uint16_t>(constant_dims, bias_data_fp16,
-                                                input_data_type);
+        std::vector<Float16> bias_data_fp16 = Float16FromFloat32(bias_data);
```
Same here.

huningxin commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -266,6 +298,13 @@ std::string GraphBuilderOrt::CreateInitializer(base::span<const uint32_t> shape,
+  requires internal::IsSupportedTensorType<DataType>
+std::string GraphBuilderOrt::CreateScalarInitializer(const DataType& value) {
+  return CreateInitializer(
+      /*shape=*/{}, base::span(&value, 1u));
```
Use `span_from_ref`?
```suggestion
      /*shape=*/{}, base::span_from_ref(value));
```

huningxin commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -241,14 +272,15 @@ std::string GraphBuilderOrt::GetOperandName(uint64_t operand_id) {
+  requires internal::IsSupportedTensorType<DataType>
+std::string GraphBuilderOrt::CreateInitializer(
+    base::span<const uint32_t> shape,
+    base::span<const DataType> data) {
```
Could you please explain why it is safe to call `base::as_byte_span(data)` when `data` is `base::span<Float16>`?

mingmingtasd commented:
Addressed your comments, PTAL again, thanks!

huningxin commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -248,18 +271,22 @@ std::string GraphBuilderOrt::GetOperandName(uint64_t operand_id) {
   } else {
+    // It is safe to call `base::as_byte_span(data)` for `base::span<const
+    // Float16>`, because `Float16` is a wrapper around `uint16_t`, not a real
+    // floating point type.
```
My concern is `Float16` is a class. The `base::as_byte_span()` of a `std::vector<Float16>` might not give us the same memory layout of `std::vector<uint16_t>`. Or did I miss anything?

mingmingtasd replied:
Oh, good question! I debugged, and find that the `class Float16` type can pass the check `internal::ByteSpanConstructibleFrom<const T&>` at  https://source.chromium.org/chromium/chromium/src/+/main:base/containers/span.h;l=1656;drc=f39c57f31413abcb41d3068cfb2c7a1718003cc5?q=base::as_byte_span&ss=chromium%2Fchromium%2Fsrc, and the webnn wpt test for float16 clamp can pass correctly, it should be OK to use `base::as_byte_span()` for `Float16`.

And I just tried if I put a `std::string` member in the `class Float16`, this check will fail. Anyway, `uint16` member is OK.

huningxin replied:
`Float16` can be converted to byte span, because it has an unique object representation. See `std::has_unique_object_representations_v`:

https://source.chromium.org/chromium/chromium/src/+/main:base/containers/span.h;l=311;drc=f39c57f31413abcb41d3068cfb2c7a1718003cc5?q=base::as_byte_span&ss=chromium%2Fchromium%2Fsrc

https://en.cppreference.com/w/cpp/types/has_unique_object_representations

It doesn't guarantee `Float16`'s memory layout (may be compiler or platform dependent) is the same as `uint16_t` which is what we want.

So, I suppose we should keep using `uint16_t` for float16 data without introducing `Float16` class, WDYT?

mingmingtasd replied:
Agree to keep using uint16_t, I will update the PR, thanks!

mingmingtasd replied:
Updated, PTAL again, thanks!

huningxin commented on services/webnn/ort/graph_builder_ort.cc:
```diff
@@ -144,24 +144,48 @@ std::string MapReduceKindToOrtOpType(mojom::Reduce::Kind kind) {
+};
+
+template <>
+struct TensorTypeMap<uint16_t> {
```
Nit: add a comment explaining we use `uint16_t` to carry bits of float16.

huningxin commented on services/webnn/ort/graph_builder_ort.h:
```diff
@@ -96,8 +113,12 @@ class GraphBuilderOrt {
+  //
+  // The provided data must match `DataType` according to the ONNX operators'
+  // data type constraints. For example: Provide float data for ONNX float data
+  // type.
```
Please mention using `uint16_t` to carry bits of float16.

huningxin reviewed (approved):
LGTM with two nits

mingmingtasd commented:
Fixed nits, PTAL, thanks!

mingmingtasd commented:
> LGTM
> 
> One minor suggestion is please add commits when addressing review comments (avoid force pushing), so reviewers can easily find the delta. Thanks!

Sure, will do~
